client/components/calypso-i18n-provider/index.tsx

```diff
@@ -2,22 +2,23 @@
  * External dependencies
  */
 import * as React from 'react';
-import { setLocaleData as setWpI18nLocaleData } from '@wordpress/i18n';
-import { I18nProvider } from '@automattic/react-i18n';
+import { createI18n, setLocaleData as setWpI18nLocaleData } from '@wordpress/i18n';
+import { I18nProvider } from '@wordpress/react-i18n';
 import i18n from 'i18n-calypso';
 import { LocaleProvider, i18nDefaultLocaleSlug } from '@automattic/i18n-utils';
 
 const CalypsoI18nProvider: React.FunctionComponent = ( { children } ) => {
-	const [ localeData, setLocaleData ] = React.useState( i18n.getLocale() );
 	const [ localeSlug, setLocaleSlug ] = React.useState( i18n.getLocaleSlug() );
 
+	// Create a new @wordpress/i18n instance when the locale changes,
+	// as `setLocaleData` doesn't replace the entire locale data, but rather merges it with the existing one,
+	// which may lead to residue translations from previous locales to remain causing displaying mixed up translations.
+	const wpI18n = React.useMemo( () => createI18n( i18n.getLocale() ), [ localeSlug ] );
+
 	React.useEffect( () => {
```

Comment:
The linter will complain about a missing `wpI18n` dependency for this hook. Trying to resolve this puzzle, I arrived at this alternative implementation:
```jsx
const createValue = () => ( {
	slug: i18n.getLocaleSlug(),
	i18n: createI18n( i18n.getLocale() ),
} );

const CalypsoI18nProvider: React.FunctionComponent = ( { children } ) => {
	const [ value, setValue ] = React.useState( createValue );

	React.useEffect( () => {
		const onChange = () => {
			setValue( ( prevValue ) => {
				// always sync the global default i18n instance
				setWpI18nLocaleData( i18n.getLocale() );

				if ( prevValue.slug !== i18n.getLocaleSlug() ) {
					// Create a new @wordpress/i18n instance when the locale changes,
					// as `setLocaleData` doesn't replace the entire locale data, but rather merges it with the existing one,
					// which may lead to residue translations from previous locales to remain causing displaying mixed up translations.
					return createValue();
				}

				// merely merge the new locale data into existing instance and don't change state
				prevValue.i18n.setLocaleData( i18n.getLocale() );
				return prevValue;
			} );
		};

		i18n.on( 'change', onChange );

		return () => {
			i18n.off( 'change', onChange );
		};
	}, [] );

	return (
		<LocaleProvider localeSlug={ value.slug || i18nDefaultLocaleSlug }>
			<I18nProvider i18n={ value.i18n }>{ children }</I18nProvider>
		</LocaleProvider>
	);
};
```
It:
- creates a local `i18n` instance
- recreates it on upstream (`i18n-calypso`) locale slug change
- passes it down in `<I18nProvider>`
- synces data changes from upstream to the local `i18n` instance
- synces data changes to the global default `i18n` instance, too. Keeping the bug where a locale change would merge data from two languages together